include/deal.II/base/function_restriction.h

```diff
@@ -0,0 +1,160 @@
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#ifndef dealii_function_restriction_h
+#define dealii_function_restriction_h
+
+#include <deal.II/base/function.h>
+#include <deal.II/base/point.h>
+#include <deal.II/base/smartpointer.h>
+#include <deal.II/base/tensor.h>
+
+DEAL_II_NAMESPACE_OPEN
+
+namespace Functions
+{
+  /**
+   * This class takes a function in dim + 1 and creates a new function in one
+   * dimension lower by restricting one of the coordinates to a given value.
+   * Mathematically this corresponds to to taking a function
+   * f = f(x, y, z),
+   * a fix value, Z, and defining a new function (the restriction)
+   * g = g(x, y) = f(x, y, Z).
+   *
+   * The dim-dimensional coordinates on the restriction are ordered starting
+   * from restricted (dim + 1)-coordinate. In particular, this means that if the
+   * y-coordinate is locked to Y in 3D, the coordinates are ordered as (z, x) on
+   * the restriction:
+   * g = g(z, x) = f(x, Y, z).
+   * This is the same convention as in BoundingBox::cross_section.
+   */
+  template <int dim>
+  class CoordinateRestriction : public Function<dim>
+  {
+  public:
+    /**
+     * Constructor, takes the (dim + 1)-coordinte direction and value that the
+     * incoming function should be restricted to.
+     *
+     * A pointer to the incoming function is stored internally, so the function
+     * must have a longer lifetime than the created restriction.
+     */
+    CoordinateRestriction(const Function<dim + 1> &function,
+                          const unsigned int       direction,
+                          const double             coordinate_value);
+
+    double
+    value(const Point<dim> &point, const unsigned int component) const override;
+
+    Tensor<1, dim>
+    gradient(const Point<dim> & point,
+             const unsigned int component) const override;
+
+    SymmetricTensor<2, dim>
+    hessian(const Point<dim> & point,
+            const unsigned int component) const override;
+
+  private:
+    // The higher-dimensional function that has been restricted.
+    const SmartPointer<const Function<dim + 1>> function;
+
+    // The (dim + 1)-coordinate direction that has been restricted.
+    const unsigned int restricted_direction;
+
+    // Value of the restriced coordinate.
+    const double coordinate_value;
+  };
+
+
+
+  /**
+   * This class creates a 1-dimensional function from a dim + 1 dimension
+   * function by restriction dim of the coordinate values to a given point.
+   * Mathematically this corresponds to taking a function f = f(x, y, z), and a
+   * point (Y, Z), and defining a new function g = g(x) = f(x, Y, Z).
+   *
+   * The coordinates of the point will be expanded in the higher-dimensional
+   * functions coordinates starting from the open-direction (and wrapping
+   * around). In particular, if we restrict to a point (Z, X) and choose to keep
```

Comment:
here and in the previous class, it might be useful to show a code snippet that illustrates how one would want to do that